sympy/matrices/random.py

```diff
@@ -0,0 +1,2378 @@
+            obj[2 * i + 1, 2 * j] = -b
+            obj[2 * i + 1, 2 * j + 1] = a
+    return obj
+
+
+def regular_to_singular(mat, rank=None):
+    r""" build matrix with linear combination of matrix columns to meet rank
+
+    Explanation
+    ===========
+    Let $\mathbf{I}$ be the $n \times n$ identity matrix
+    and $\mathbf{D}$ an diagonal matrix with only zero and one entries
+    of rank $r$ (i.e. all entries but $n-r$ diagonal entries are zero).
+
+    If a full rank matrix $\mathbf{A}$ is given, then
+
+    .. math::
+
+        \mathbf{B}
+        = \mathbf{D} \cdot
+        (\mathbf{I} + \mathbf{A} \cdot (\mathbf{I}-\mathbf{D}))
+        = \mathbf{D} + \mathbf{D} \cdot \mathbf{A} \cdot(\mathbf{I}-\mathbf{D})
+
+    is of rank $r$, too.  Some for $\mathbf{A}\cdot\mathbf{B}$
+    which contains $r$ columns with $\mathbf{A}$.
+
+    If $\mathbf{A}$ is an upper triangular matrix $\mathbf{B}$
+    as well $\mathbf{A} \cdot \mathbf{B}$ is one, too.
+
+    Here, $\mathbf{A} \cdot \mathbf{B}$ is returned.
+    Note, if $n=r$ the matrix $\mathbf{B}$ is the identity.
+
+    Examples
+    ========
+    >>> from sympy import Matrix
+    >>> from sympy.matrices.random import regular_to_singular
+    >>> m = Matrix([[1, 2, 3],[4, 5, 6], [7, 8 ,0]])
+    >>> m.rank()
+    3
+    >>> n = regular_to_singular(m, 2)
+    >>> n
+    Matrix([
+    [1, 2, 15],
+    [4, 5, 42],
+    [7, 8, 69]])
+    >>> n.rank()
+    2
+
+    Parameters
+    ==========
+    mat : Matrix
+        a complex matrix
+    rank : integer
+        rank of matrix
+
+    """
+    if rank is None:
+        return mat
+    dim = _shape(mat)[0]
+    if rank == dim:
+        return mat
+    i = _eye(dim)
+    p = permutation(dim)
+    d = projection(dim, (0, rank))
+    d = _multiply(p.inv(), d, p)
+    md = _multiply(mat, d)
+    return md + _multiply(md, mat, i - d)
+
+
+# === base matrices ===
+
+
+def projection(dim,
+               index=None):
+    r"""A randomly generated n x n projection matrix
+
+    Explanation
+    ===========
+
+    A projection is a identity like matrix
+    but with zero diagonal entiries off *index*.
+
+    Examples
+    ========
+
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import projection
+    >>> seed(1)
+
+    >>> projection(3)
+    Matrix([
+    [0, 0, 0],
+    [0, 1, 0],
+    [0, 0, 0]])
+
+    >>> projection(3, index=(1,3))  # no more random
+    Matrix([
+    [0, 0, 0],
+    [0, 1, 0],
+    [0, 0, 1]])
+
+    Parameters
+    ==========
+
+    dim : integer
+        dimension n of matrix
+    index : tuple(integer, integer) (optional)
+        coordinates ``(i,j)`` for start (included) and end (excluded)
+        to project onto
+
+    See Also
+    ========
+    jordan
+    elementary
+    diagonal_normal
+
+    """
+
+    index = index or _sample(range(dim + 1), 2)
+    obj = _eye(dim)
+    start, end = sorted(index)
+    for i in range(dim):
+        v = 1 if start <= i < end else 0
+        obj[i, i] = v
```

Comment:
Why not just loop like `for i in range(start, end)`?

On the other hand it would seem more natural to me for this function to take a list of indices rather than a start stop tuple.